### ISSUE TITLE
About the performance

### ISSUE THREAD

**Issue by RunDevil-Run** [Visual-AI/SPTNet#13]
Hello, I have tried to train the model using simgcd pretrained weights on CUB. SCARS, FGVC, but the performance I get from the model is almost the same as the pretrained simgcd weights, I wonder if there are any details I overlooked? I run the training using exactly the provided command.

**Comment by whj363636** [comment]
Hi @RunDevil-Run 

Could you provide the number of your pretrained SimGCD model on CUB, SCARS and FGVC? Did you train the SimGCD by yourself following their repo, _or just download the models from the link I provided in the readme file_? Because the models provided in the link **are exactly the best SPTNet models** rather than SimGCD itself, it is not strange to get the same results to so-called ``SimGCD''. 

If you indeed obtain the pretrained SimGCD model by yourself and the number is normal, have you evaluated the model on validation dataset through training and pick the best one or just select the last one? In our paper, we pick the best one on validation dataset.

**Comment by RunDevil-Run** [comment]
Sorry to bother you again, I tried to run the training several times, and the best epoch ACC indeed the same as paper. But I noticed that in both train_spt.py and eval.py, the evaluation is based on **test_loader_unlabelled**, I wonder if this is exactly the validation set you mentioned or I have to modify the code to use the **test_loader_labelled** as the validation set?
Since in the original simgcd repository, they suggest to directly use the last epoch as the final result, but in their commit history, they indeed used model selection to select best model from validation set, which is **test_loader_labelled** in their code history.
So I wonder the correct procedure for this task is to evaluate on **test_loader_labelled** (labeled class from training set but not overlap to training set) and pick best one, then use **test_loader_unlabelled** to get the final accuracy. Is my understanding correct? 
Thank you very much!

**Comment by whj363636** [comment]
Hi @RunDevil-Run 

Great to know you get there :). And yes, you're right. We pick the best hyperparameters evaluated on labelled data. The evaluation part you mentioned in train_spt.py is intended for a different project (shared with the same codebase before refactoring), as indicated by the comment :). I have made the necessary adjustments to prevent any confusion. Thank you for bringing that to my attention.

**Comment by RunDevil-Run** [comment]
Sorry to bother you again, I made a mistake in my last comment. I recheck the logs and here are my results on three datasets after training.

CUB: All: 0.6050 Old: 0.6744 New: 0.5702
SCARS: All: 0.5545 Old: 0.7271 New: 0.4712
FGVC: All: 0.5645 Old: 0.6417 New: 0.5259

These numbers are stable during 5 independent training respectively, and I got them by directly print https://github.com/Visual-AI/SPTNet/blob/86a65e3dc6c4100855947017c0cfc22293c68805/train_spt.py#L314 
through loguru logger tool, as same as simgcd, and change **test_loader_labelled** to **test_loader_unlabelled** . I strictly follow the training configuration in table7 from appendixB for different datasets, and my pytorch version is 1.12.0 with cuda 11.6, torchvision 0.13.0.

The only modification I made is adding logger tool to the original codes, could you give me any suggestion? Thank you for your reply!

**Comment by whj363636** [comment]
Hi @RunDevil-Run 

Sorry for late reply, I have spent time in re-running the code directly downloaded from GitHub, both the original version and the code after refactoring, using the recommended environment specified in requirements.txt (torch==1.13.1, torchvision==0.14.1) with CUDA 11.6. I have included the logs below for comparison purposes:
 
[aircraft.log](https://github.com/Visual-AI/SPTNet/files/15124663/aircraft.log)
[aircraft_github.log](https://github.com/Visual-AI/SPTNet/files/15124664/aircraft_github.log)

I followed the steps you outlined, which involved changing test_loader_labelled to test_loader_unlabelled, and print results for each epoch. However, I did not observe any noticeable performance difference between these two versions, including the comparison with the paper's results. Meanwhile, I would like to clarify that I noticed the accuracy you provided for FGVC is already higher than SimGCD (although there is still a discrepancy compared to the results in our paper), rather than being "as same as SimGCD".
